python/ray/data/_internal/logical/rules/operator_fusion.py

```diff
@@ -93,19 +93,33 @@ def _fuse_all_to_all_operators_in_dag(
     ) -> AllToAllOperator:
         """Starting at the given operator, traverses up the DAG of operators
         and recursively fuses compatible MapOperator -> AllToAllOperator pairs.
+
+        Also, sets the target block size of the immediately upstream map op to
+        match the shuffle block size. We use a larger block size for shuffles
+        because tiny blocks are bad for I/O performance.
+
         Returns the current (root) operator after completing upstream operator fusions.
         """
         upstream_ops = dag.input_dependencies
         while (
             len(upstream_ops) == 1
             and isinstance(dag, AllToAllOperator)
             and isinstance(upstream_ops[0], MapOperator)
-            and self._can_fuse(dag, upstream_ops[0])
         ):
-            # TODO(swang): Set block size for Sort and Aggregate.
-            # Fuse operator with its upstream op.
-            dag = self._get_fused_all_to_all_operator(dag, upstream_ops[0])
-            upstream_ops = dag.input_dependencies
+            if self._can_fuse(dag, upstream_ops[0]):
+                # Fuse operator with its upstream op.
+                dag = self._get_fused_all_to_all_operator(dag, upstream_ops[0])
+                upstream_ops = dag.input_dependencies
+            else:
+                # Propagate target max block size to the upstream map op. This
+                # is necessary even when fusion is not allowed, so that the map
+                # op will produce the right block size for the shuffle op to
+                # consume.
+                map_op = upstream_ops[0]
+                map_op._target_max_block_size = self._get_merged_target_max_block_size(
+                    upstream_ops[0].target_max_block_size, dag.target_max_block_size
+                )
+                break
```

Comment:
I'm wondering if it makes more sense to have a separate rule for propagating the block size, since these 2 ops are not fused. But I don't have strong preference.

Reply:
Seems a bit safer to me to put them in the same rule since the order of rule application matters (we should always run this after applying fusion).